xds/src/main/java/io/grpc/xds/XdsClusterResource.java

```diff
@@ -169,9 +174,48 @@ static CdsUpdate processCluster(Cluster cluster,
     updateBuilder.filterMetadata(
         ImmutableMap.copyOf(cluster.getMetadata().getFilterMetadataMap()));
 
+    try {
+      ImmutableMap<String, Object> parsedFilterMetadata =
+          parseClusterMetadata(cluster.getMetadata());
+      updateBuilder.parsedMetadata(parsedFilterMetadata);
+    } catch (InvalidProtocolBufferException e) {
+      throw new ResourceInvalidException(
+          "Failed to parse xDS filter metadata for cluster '" + cluster.getName() + "': "
+              + e.getMessage(), e);
+    }
+
     return updateBuilder.build();
   }
 
+  private static ImmutableMap<String, Object> parseClusterMetadata(Metadata metadata)
+      throws InvalidProtocolBufferException {
+    ImmutableMap.Builder<String, Object> parsedMetadata = ImmutableMap.builder();
+
+    ClusterMetadataRegistry registry = ClusterMetadataRegistry.getInstance();
+    // Process typed_filter_metadata
+    for (Map.Entry<String, Any> entry : metadata.getTypedFilterMetadataMap().entrySet()) {
+      String key = entry.getKey();
+      Any value = entry.getValue();
+      ClusterMetadataValueParser parser = registry.findParser(value.getTypeUrl());
+      if (parser != null) {
+        Object parsedValue = parser.parse(value);
+        parsedMetadata.put(key, parsedValue);
+      }
+    }
+
+    // Process filter_metadata for remaining keys
+    for (Map.Entry<String, Struct> entry : metadata.getFilterMetadataMap().entrySet()) {
+      String key = entry.getKey();
+      if (!parsedMetadata.build().containsKey(key)) {
```

Comment:
You are build()ing a map for every key? Surely this isn't what we want.

Reply:
Oh thats the performance issue.
Alright I have one idea which tweaks the [gRFC saying](https://github.com/grpc/proposal/blob/master/A83-xds-gcp-authn-filter.md#xds-cluster-metadata),

>The logic to validate cluster metadata will look something like this (pseudo-code):
```
parsed_metadata = {}  # Value is either JSON or parsed object
# First process typed_filter_metadata.
for key, any_field in cluster_metadata.typed_filter_metadata.items():
  parser = metadata_registry.FindParser(any_field.type_url)
  if parser is not None:
    value = parser.Parse(any_field.value)
    if value is None:
      return NACK  # Parsing failed, reject resource
    parsed_metadata[key] = value
# Now process filter_metadata.  We look only at keys that were not
# already added from typed_filter_metadata.
for key, struct_field in cluster_metadata.filter_metadata.items():
  if key not in parsed_metadata:
    parsed_metadata[key] = ConvertToJson(struct_field)
```

In here, if we process filter_metadata before typed_filter_metadata then we do not need to ask if containsKey at all. We just replace whatever there is already in the map with the new value. So like this, if the key is totally new, it gets added and if the key is already there, then it updates the value.

Sounds good?

The other way could be using additional HashSet of Keys and checking if we ever see this key before or not!

Reply:
Can can do `parsedMetadata.build()` once before the loop, and reuse it for each check.

I don't think HashSet would be the next way. I'd just use a HashMap instead of ImmutableMap.Builder, and convert it with `ImmutableMap.copyOf()` when returning.

Reordering the loops could be fine. The part to be wary of is validation errors; we wouldn't want to _begin_ failing because now we are processing a key we ordinarily wouldn't. Right now though, the JSON conversion essentially can't fail; it could if JSON itself changed to support a new type, but that's essentially impossible.